src/backend/distributed/utils/enable_ssl.c

```diff
@@ -0,0 +1,444 @@
+ * if they are not already installed in the configured location.
+ */
+Datum
+citus_setup_ssl(PG_FUNCTION_ARGS)
+{
+#ifndef USE_SSL
+	ereport(WARNING, (errmsg("can not setup ssl on postgres that is not compiled with "
+							 "ssl support")));
+#else /* USE_SSL */
+	if (!EnableSSL && ShouldUseAutoSSL())
+	{
+		Node *enableSSLParseTree = NULL;
+
+		ereport(LOG, (errmsg("citus extension created on postgres without ssl enabled, "
+							 "turning it on during creation of the extension")));
+
+		/* execute the alter system statement to enable ssl on within postgres */
+		enableSSLParseTree = ParseTreeNode(ENABLE_SSL_QUERY);
+		AlterSystemSetConfigFile((AlterSystemStmt *) enableSSLParseTree);
+
+		/*
+		 * ssl=on requires that a key and certificate are present, since we have
+		 * enabled ssl mode here chances are the user didn't install credentials already.
+		 *
+		 * This function will check if they are available and if not it will generate a
+		 * self singed certificate.
+		 */
+		CreateCertificatesWhenNeeded();
+
+#if PG_VERSION_NUM >= 100000
+		GloballyReloadConfig();
+#else /* PG_VERSION_NUM < 100000 */
+		ereport(WARNING, (errmsg("restart of postgres required"),
+						  errdetail("citus enables ssl in postgres. Postgres "
+									"versions before 10.0 require a restart for changes "
+									"to ssl to take effect."),
+						  errhint("when restarting is not possible disable ssl and "
+								  "change citus.conn_nodeinfo to have sslmode lower "
+								  "then require.")));
+#endif
+	}
+#endif /* USE_SSL */
+
+	PG_RETURN_NULL();
+}
+
+
+/*
+ * citus_check_defaults_for_sslmode is called in the extension upgrade path when
+ * users upgrade from a previous version to a version that has ssl enabled by default, and
+ * only when the changed default value conflicts with the setup of the user.
+ *
+ * Once it is determined that the default value for citus.node_conninfo is used verbatim
+ * with ssl not enabled on the cluster it will reinstate the old default value for
+ * citus.node_conninfo.
+ *
+ * In effect this is to not impose the overhead of ssl on an already existing cluster that
+ * didn't have it enabled already.
+ */
+Datum
+citus_check_defaults_for_sslmode(PG_FUNCTION_ARGS)
+{
+	bool configChanged = false;
+
+	if (EnableSSL)
+	{
+		/* since ssl is on we do not have to change any sslmode back to prefer */
+		PG_RETURN_NULL();
+	}
+
+	if (strcmp(NodeConninfo, "sslmode=require") == 0)
+	{
+		Node *resetCitusNodeConnInfoParseTree = NULL;
+
+		/* execute the alter system statement to reset node_conninfo to the old default */
+
+		ereport(LOG, (errmsg("reset citus.node_conninfo to old default value as the new "
+							 "value is incompatible with the current ssl setting")));
+
+		resetCitusNodeConnInfoParseTree = ParseTreeNode(RESET_CITUS_NODE_CONNINFO);
+		AlterSystemSetConfigFile((AlterSystemStmt *) resetCitusNodeConnInfoParseTree);
+		configChanged = true;
+	}
+
+	if (configChanged)
+	{
+		GloballyReloadConfig();
```

Comment:
This is a bit weird, why isn't it called in the block above?

Reply:
This is to prepare enterprise to have an extra block in between that possibly can also change configuration.

Other solution could be to reload the global config potentially twice.